// @angular/material/esm2015/expansion/accordion.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, Input, ContentChildren, QueryList } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { CdkAccordion } from '@angular/cdk/accordion';
import { FocusKeyManager } from '@angular/cdk/a11y';
import { HOME, END, hasModifierKey } from '@angular/cdk/keycodes';
import { startWith } from 'rxjs/operators';
import { MAT_ACCORDION } from './accordion-base';
import { MatExpansionPanelHeader } from './expansion-panel-header';
/**
 * Directive for a Material Design Accordion.
 */
let MatAccordion = /** @class */ (() => {
    class MatAccordion extends CdkAccordion {
        constructor() {
            super(...arguments);
            /** Headers belonging to this accordion. */
            this._ownHeaders = new QueryList();
            this._hideToggle = false;
            /**
             * Display mode used for all expansion panels in the accordion. Currently two display
             * modes exist:
             *  default - a gutter-like spacing is placed around any expanded panel, placing the expanded
             *     panel at a different elevation from the rest of the accordion.
             *  flat - no spacing is placed around expanded panels, showing all panels at the same
             *     elevation.
             */
            this.displayMode = 'default';
            /** The position of the expansion indicator. */
            this.togglePosition = 'after';
        }
        /** Whether the expansion indicator should be hidden. */
        get hideToggle() { return this._hideToggle; }
        set hideToggle(show) { this._hideToggle = coerceBooleanProperty(show); }
        ngAfterContentInit() {
            this._headers.changes
                .pipe(startWith(this._headers))
                .subscribe((headers) => {
                this._ownHeaders.reset(headers.filter(header => header.panel.accordion === this));
                this._ownHeaders.notifyOnChanges();
            });
            this._keyManager = new FocusKeyManager(this._ownHeaders).withWrap();
        }
        /** Handles keyboard events coming in from the panel headers. */
        _handleHeaderKeydown(event) {
            const { keyCode } = event;
            const manager = this._keyManager;
            if (keyCode === HOME) {
                if (!hasModifierKey(event)) {
                    manager.setFirstItemActive();
                    event.preventDefault();
                }
            }
            else if (keyCode === END) {
                if (!hasModifierKey(event)) {
                    manager.setLastItemActive();
                    event.preventDefault();
                }
            }
            else {
                this._keyManager.onKeydown(event);
            }
        }
        _handleHeaderFocus(header) {
            this._keyManager.updateActiveItem(header);
        }
    }
    MatAccordion.decorators = [
        { type: Directive, args: [{
                    selector: 'mat-accordion',
                    exportAs: 'matAccordion',
                    inputs: ['multi'],
                    providers: [{
                            provide: MAT_ACCORDION,
                            useExisting: MatAccordion
                        }],
                    host: {
                        class: 'mat-accordion',
                        // Class binding which is only used by the test harness as there is no other
                        // way for the harness to detect if multiple panel support is enabled.
                        '[class.mat-accordion-multi]': 'this.multi',
                    }
                },] }
    ];
    MatAccordion.propDecorators = {
        _headers: [{ type: ContentChildren, args: [MatExpansionPanelHeader, { descendants: true },] }],
        hideToggle: [{ type: Input }],
        displayMode: [{ type: Input }],
        togglePosition: [{ type: Input }]
    };
    return MatAccordion;
})();
export { MatAccordion };
