// @angular/material/esm2015/list/testing/nav-list-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { HarnessPredicate } from '@angular/cdk/testing';
import { MatListHarnessBase } from './list-harness-base';
import { getListItemPredicate, MatListItemHarnessBase } from './list-item-harness-base';
/** Harness for interacting with a standard mat-nav-list in tests. */
let MatNavListHarness = /** @class */ (() => {
    class MatNavListHarness extends MatListHarnessBase {
        constructor() {
            super(...arguments);
            this._itemHarness = MatNavListItemHarness;
        }
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatNavListHarness` that meets
         * certain criteria.
         * @param options Options for filtering which nav list instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatNavListHarness, options);
        }
    }
    /** The selector for the host element of a `MatNavList` instance. */
    MatNavListHarness.hostSelector = 'mat-nav-list';
    return MatNavListHarness;
})();
export { MatNavListHarness };
/** Harness for interacting with a nav list item. */
let MatNavListItemHarness = /** @class */ (() => {
    class MatNavListItemHarness extends MatListItemHarnessBase {
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatNavListItemHarness` that
         * meets certain criteria.
         * @param options Options for filtering which nav list item instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return getListItemPredicate(MatNavListItemHarness, options)
                .addOption('href', options.href, (harness, href) => __awaiter(this, void 0, void 0, function* () { return HarnessPredicate.stringMatches(harness.getHref(), href); }));
        }
        /** Gets the href for this nav list item. */
        getHref() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).getAttribute('href');
            });
        }
        /** Clicks on the nav list item. */
        click() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).click();
            });
        }
        /** Focuses the nav list item. */
        focus() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).focus();
            });
        }
        /** Blurs the nav list item. */
        blur() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).blur();
            });
        }
    }
    /** The selector for the host element of a `MatListItem` instance. */
    MatNavListItemHarness.hostSelector = ['mat-list-item', 'a[mat-list-item]', 'button[mat-list-item]']
        .map(selector => `${MatNavListHarness.hostSelector} ${selector}`)
        .join(',');
    return MatNavListItemHarness;
})();
export { MatNavListItemHarness };
