// @angular/material/esm2015/list/testing/selection-list-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { HarnessPredicate } from '@angular/cdk/testing';
import { MatListHarnessBase } from './list-harness-base';
import { getListItemPredicate, MatListItemHarnessBase } from './list-item-harness-base';
/** Harness for interacting with a standard mat-selection-list in tests. */
let MatSelectionListHarness = /** @class */ (() => {
    class MatSelectionListHarness extends MatListHarnessBase {
        constructor() {
            super(...arguments);
            this._itemHarness = MatListOptionHarness;
        }
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatSelectionListHarness` that meets
         * certain criteria.
         * @param options Options for filtering which selection list instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatSelectionListHarness, options);
        }
        /** Whether the selection list is disabled. */
        isDisabled() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield (yield this.host()).getAttribute('aria-disabled')) === 'true';
            });
        }
        /**
         * Selects all items matching any of the given filters.
         * @param filters Filters that specify which items should be selected.
         */
        selectItems(...filters) {
            return __awaiter(this, void 0, void 0, function* () {
                const items = yield this._getItems(filters);
                yield Promise.all(items.map(item => item.select()));
            });
        }
        /**
         * Deselects all items matching any of the given filters.
         * @param filters Filters that specify which items should be deselected.
         */
        deselectItems(...filters) {
            return __awaiter(this, void 0, void 0, function* () {
                const items = yield this._getItems(filters);
                yield Promise.all(items.map(item => item.deselect()));
            });
        }
        /** Gets all items matching the given list of filters. */
        _getItems(filters) {
            return __awaiter(this, void 0, void 0, function* () {
                if (!filters.length) {
                    return this.getItems();
                }
                return [].concat(...yield Promise.all(filters.map(filter => this.locatorForAll(MatListOptionHarness.with(filter))())));
            });
        }
    }
    /** The selector for the host element of a `MatSelectionList` instance. */
    MatSelectionListHarness.hostSelector = 'mat-selection-list';
    return MatSelectionListHarness;
})();
export { MatSelectionListHarness };
/** Harness for interacting with a list option. */
let MatListOptionHarness = /** @class */ (() => {
    class MatListOptionHarness extends MatListItemHarnessBase {
        constructor() {
            super(...arguments);
            this._itemContent = this.locatorFor('.mat-list-item-content');
        }
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatListOptionHarness` that
         * meets certain criteria.
         * @param options Options for filtering which list option instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return getListItemPredicate(MatListOptionHarness, options)
                .addOption('is selected', options.selected, (harness, selected) => __awaiter(this, void 0, void 0, function* () { return (yield harness.isSelected()) === selected; }));
        }
        /** Gets the position of the checkbox relative to the list option content. */
        getCheckboxPosition() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield (yield this._itemContent()).hasClass('mat-list-item-content-reverse')) ?
                    'after' : 'before';
            });
        }
        /** Whether the list option is selected. */
        isSelected() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield (yield this.host()).getAttribute('aria-selected')) === 'true';
            });
        }
        /** Whether the list option is disabled. */
        isDisabled() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield (yield this.host()).getAttribute('aria-disabled')) === 'true';
            });
        }
        /** Focuses the list option. */
        focus() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).focus();
            });
        }
        /** Blurs the list option. */
        blur() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).blur();
            });
        }
        /** Toggles the checked state of the checkbox. */
        toggle() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).click();
            });
        }
        /**
         * Puts the list option in a checked state by toggling it if it is currently unchecked, or doing
         * nothing if it is already checked.
         */
        select() {
            return __awaiter(this, void 0, void 0, function* () {
                if (!(yield this.isSelected())) {
                    return this.toggle();
                }
            });
        }
        /**
         * Puts the list option in an unchecked state by toggling it if it is currently checked, or doing
         * nothing if it is already unchecked.
         */
        deselect() {
            return __awaiter(this, void 0, void 0, function* () {
                if (yield this.isSelected()) {
                    return this.toggle();
                }
            });
        }
    }
    /** The selector for the host element of a `MatListOption` instance. */
    MatListOptionHarness.hostSelector = 'mat-list-option';
    return MatListOptionHarness;
})();
export { MatListOptionHarness };
