// @angular/material/esm2015/button-toggle/testing/button-toggle-group-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { ComponentHarness, HarnessPredicate } from '@angular/cdk/testing';
import { MatButtonToggleHarness } from './button-toggle-harness';
/** Harness for interacting with a standard mat-button-toggle in tests. */
let MatButtonToggleGroupHarness = /** @class */ (() => {
    class MatButtonToggleGroupHarness extends ComponentHarness {
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatButtonToggleGroupHarness`
         * that meets certain criteria.
         * @param options Options for filtering which button toggle instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatButtonToggleGroupHarness, options);
        }
        /**
         * Gets the button toggles that are inside the group.
         * @param filter Optionally filters which toggles are included.
         */
        getToggles(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatButtonToggleHarness.with(filter))();
            });
        }
        /** Gets whether the button toggle group is disabled. */
        isDisabled() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield (yield this.host()).getAttribute('aria-disabled')) === 'true';
            });
        }
        /** Gets whether the button toggle group is laid out vertically. */
        isVertical() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this.host()).hasClass('mat-button-toggle-vertical');
            });
        }
        /** Gets the appearance that the group is using. */
        getAppearance() {
            return __awaiter(this, void 0, void 0, function* () {
                const host = yield this.host();
                const className = 'mat-button-toggle-group-appearance-standard';
                return (yield host.hasClass(className)) ? 'standard' : 'legacy';
            });
        }
    }
    /** The selector for the host element of a `MatButton` instance. */
    MatButtonToggleGroupHarness.hostSelector = 'mat-button-toggle-group';
    return MatButtonToggleGroupHarness;
})();
export { MatButtonToggleGroupHarness };
