// @angular/material/esm2015/table/testing/row-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { ComponentHarness, HarnessPredicate } from '@angular/cdk/testing';
import { MatCellHarness, MatHeaderCellHarness, MatFooterCellHarness } from './cell-harness';
/** Harness for interacting with a standard Angular Material table row. */
let MatRowHarness = /** @class */ (() => {
    class MatRowHarness extends ComponentHarness {
        /**
         * Gets a `HarnessPredicate` that can be used to search for a table row with specific attributes.
         * @param options Options for narrowing the search
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatRowHarness, options);
        }
        /** Gets a list of `MatCellHarness` for all cells in the row. */
        getCells(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatCellHarness.with(filter))();
            });
        }
        /** Gets the text of the cells in the row. */
        getCellTextByIndex(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByIndex(this, filter);
            });
        }
        /** Gets the text inside the row organized by columns. */
        getCellTextByColumnName() {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByColumnName(this);
            });
        }
    }
    /** The selector for the host element of a `MatRowHarness` instance. */
    MatRowHarness.hostSelector = '.mat-row';
    return MatRowHarness;
})();
export { MatRowHarness };
/** Harness for interacting with a standard Angular Material table header row. */
let MatHeaderRowHarness = /** @class */ (() => {
    class MatHeaderRowHarness extends ComponentHarness {
        /**
         * Gets a `HarnessPredicate` that can be used to search for
         * a table header row with specific attributes.
         * @param options Options for narrowing the search
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatHeaderRowHarness, options);
        }
        /** Gets a list of `MatHeaderCellHarness` for all cells in the row. */
        getCells(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatHeaderCellHarness.with(filter))();
            });
        }
        /** Gets the text of the cells in the header row. */
        getCellTextByIndex(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByIndex(this, filter);
            });
        }
        /** Gets the text inside the header row organized by columns. */
        getCellTextByColumnName() {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByColumnName(this);
            });
        }
    }
    /** The selector for the host element of a `MatHeaderRowHarness` instance. */
    MatHeaderRowHarness.hostSelector = '.mat-header-row';
    return MatHeaderRowHarness;
})();
export { MatHeaderRowHarness };
/** Harness for interacting with a standard Angular Material table footer row. */
let MatFooterRowHarness = /** @class */ (() => {
    class MatFooterRowHarness extends ComponentHarness {
        /**
         * Gets a `HarnessPredicate` that can be used to search for
         * a table footer row cell with specific attributes.
         * @param options Options for narrowing the search
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatFooterRowHarness, options);
        }
        /** Gets a list of `MatFooterCellHarness` for all cells in the row. */
        getCells(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatFooterCellHarness.with(filter))();
            });
        }
        /** Gets the text of the cells in the footer row. */
        getCellTextByIndex(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByIndex(this, filter);
            });
        }
        /** Gets the text inside the footer row organized by columns. */
        getCellTextByColumnName() {
            return __awaiter(this, void 0, void 0, function* () {
                return getCellTextByColumnName(this);
            });
        }
    }
    /** The selector for the host element of a `MatFooterRowHarness` instance. */
    MatFooterRowHarness.hostSelector = '.mat-footer-row';
    return MatFooterRowHarness;
})();
export { MatFooterRowHarness };
function getCellTextByIndex(harness, filter) {
    return __awaiter(this, void 0, void 0, function* () {
        const cells = yield harness.getCells(filter);
        return Promise.all(cells.map(cell => cell.getText()));
    });
}
function getCellTextByColumnName(harness) {
    return __awaiter(this, void 0, void 0, function* () {
        const output = {};
        const cells = yield harness.getCells();
        const cellsData = yield Promise.all(cells.map(cell => {
            return Promise.all([cell.getColumnName(), cell.getText()]);
        }));
        cellsData.forEach(([columnName, text]) => output[columnName] = text);
        return output;
    });
}
