// @angular/material/esm2015/table/testing/table-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { ComponentHarness, HarnessPredicate } from '@angular/cdk/testing';
import { MatRowHarness, MatHeaderRowHarness, MatFooterRowHarness, } from './row-harness';
/** Harness for interacting with a standard mat-table in tests. */
let MatTableHarness = /** @class */ (() => {
    class MatTableHarness extends ComponentHarness {
        /**
         * Gets a `HarnessPredicate` that can be used to search for a table with specific attributes.
         * @param options Options for narrowing the search
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatTableHarness, options);
        }
        /** Gets all of the header rows in a table. */
        getHeaderRows(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatHeaderRowHarness.with(filter))();
            });
        }
        /** Gets all of the regular data rows in a table. */
        getRows(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatRowHarness.with(filter))();
            });
        }
        /** Gets all of the footer rows in a table. */
        getFooterRows(filter = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                return this.locatorForAll(MatFooterRowHarness.with(filter))();
            });
        }
        /** Gets the text inside the entire table organized by rows. */
        getCellTextByIndex() {
            return __awaiter(this, void 0, void 0, function* () {
                const rows = yield this.getRows();
                return Promise.all(rows.map(row => row.getCellTextByIndex()));
            });
        }
        /** Gets the text inside the entire table organized by columns. */
        getCellTextByColumnName() {
            return __awaiter(this, void 0, void 0, function* () {
                const [headerRows, footerRows, dataRows] = yield Promise.all([
                    this.getHeaderRows(),
                    this.getFooterRows(),
                    this.getRows()
                ]);
                const text = {};
                const [headerData, footerData, rowsData] = yield Promise.all([
                    Promise.all(headerRows.map(row => row.getCellTextByColumnName())),
                    Promise.all(footerRows.map(row => row.getCellTextByColumnName())),
                    Promise.all(dataRows.map(row => row.getCellTextByColumnName())),
                ]);
                rowsData.forEach(data => {
                    Object.keys(data).forEach(columnName => {
                        const cellText = data[columnName];
                        if (!text[columnName]) {
                            text[columnName] = {
                                headerText: getCellTextsByColumn(headerData, columnName),
                                footerText: getCellTextsByColumn(footerData, columnName),
                                text: []
                            };
                        }
                        text[columnName].text.push(cellText);
                    });
                });
                return text;
            });
        }
    }
    /** The selector for the host element of a `MatTableHarness` instance. */
    MatTableHarness.hostSelector = '.mat-table';
    return MatTableHarness;
})();
export { MatTableHarness };
/** Extracts the text of cells only under a particular column. */
function getCellTextsByColumn(rowsData, column) {
    const columnTexts = [];
    rowsData.forEach(data => {
        Object.keys(data).forEach(columnName => {
            if (columnName === column) {
                columnTexts.push(data[columnName]);
            }
        });
    });
    return columnTexts;
}
