// @angular/material/esm2015/grid-list/testing/grid-tile-harness.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter } from "tslib";
import { ComponentHarness, HarnessPredicate } from '@angular/cdk/testing';
/** Harness for interacting with a standard `MatGridTitle` in tests. */
let MatGridTileHarness = /** @class */ (() => {
    class MatGridTileHarness extends ComponentHarness {
        constructor() {
            super(...arguments);
            this._header = this.locatorForOptional('.mat-grid-tile-header');
            this._footer = this.locatorForOptional('.mat-grid-tile-footer');
            this._avatar = this.locatorForOptional('.mat-grid-avatar');
        }
        /**
         * Gets a `HarnessPredicate` that can be used to search for a `MatGridTileHarness`
         * that meets certain criteria.
         * @param options Options for filtering which dialog instances are considered a match.
         * @return a `HarnessPredicate` configured with the given options.
         */
        static with(options = {}) {
            return new HarnessPredicate(MatGridTileHarness, options)
                .addOption('headerText', options.headerText, (harness, pattern) => HarnessPredicate.stringMatches(harness.getHeaderText(), pattern))
                .addOption('footerText', options.footerText, (harness, pattern) => HarnessPredicate.stringMatches(harness.getFooterText(), pattern));
        }
        /** Gets the amount of rows that the grid-tile takes up. */
        getRowspan() {
            return __awaiter(this, void 0, void 0, function* () {
                return Number(yield (yield this.host()).getAttribute('rowspan'));
            });
        }
        /** Gets the amount of columns that the grid-tile takes up. */
        getColspan() {
            return __awaiter(this, void 0, void 0, function* () {
                return Number(yield (yield this.host()).getAttribute('colspan'));
            });
        }
        /** Whether the grid-tile has a header. */
        hasHeader() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this._header()) !== null;
            });
        }
        /** Whether the grid-tile has a footer. */
        hasFooter() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this._footer()) !== null;
            });
        }
        /** Whether the grid-tile has an avatar. */
        hasAvatar() {
            return __awaiter(this, void 0, void 0, function* () {
                return (yield this._avatar()) !== null;
            });
        }
        /** Gets the text of the header if present. */
        getHeaderText() {
            return __awaiter(this, void 0, void 0, function* () {
                // For performance reasons, we do not use "hasHeader" as
                // we would then need to query twice for the header.
                const headerEl = yield this._header();
                return headerEl ? headerEl.text() : null;
            });
        }
        /** Gets the text of the footer if present. */
        getFooterText() {
            return __awaiter(this, void 0, void 0, function* () {
                // For performance reasons, we do not use "hasFooter" as
                // we would then need to query twice for the footer.
                const headerEl = yield this._footer();
                return headerEl ? headerEl.text() : null;
            });
        }
    }
    /** The selector for the host element of a `MatGridTile` instance. */
    MatGridTileHarness.hostSelector = '.mat-grid-tile';
    return MatGridTileHarness;
})();
export { MatGridTileHarness };
