// @angular/material/esm2015/input/autosize.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { CdkTextareaAutosize } from '@angular/cdk/text-field';
import { Directive, Input } from '@angular/core';
/**
 * Directive to automatically resize a textarea to fit its content.
 * @deprecated Use `cdkTextareaAutosize` from `@angular/cdk/text-field` instead.
 * @breaking-change 8.0.0
 */
let MatTextareaAutosize = /** @class */ (() => {
    class MatTextareaAutosize extends CdkTextareaAutosize {
        get matAutosizeMinRows() { return this.minRows; }
        set matAutosizeMinRows(value) { this.minRows = value; }
        get matAutosizeMaxRows() { return this.maxRows; }
        set matAutosizeMaxRows(value) { this.maxRows = value; }
        get matAutosize() { return this.enabled; }
        set matAutosize(value) { this.enabled = value; }
        get matTextareaAutosize() { return this.enabled; }
        set matTextareaAutosize(value) { this.enabled = value; }
    }
    MatTextareaAutosize.decorators = [
        { type: Directive, args: [{
                    selector: 'textarea[mat-autosize], textarea[matTextareaAutosize]',
                    exportAs: 'matTextareaAutosize',
                    inputs: ['cdkAutosizeMinRows', 'cdkAutosizeMaxRows'],
                    host: {
                        'class': 'cdk-textarea-autosize mat-autosize',
                        // Textarea elements that have the directive applied should have a single row by default.
                        // Browsers normally show two rows by default and therefore this limits the minRows binding.
                        'rows': '1',
                    },
                },] }
    ];
    MatTextareaAutosize.propDecorators = {
        matAutosizeMinRows: [{ type: Input }],
        matAutosizeMaxRows: [{ type: Input }],
        matAutosize: [{ type: Input, args: ['mat-autosize',] }],
        matTextareaAutosize: [{ type: Input }]
    };
    return MatTextareaAutosize;
})();
export { MatTextareaAutosize };
